Shell: String to numeric conversion utils

Andrewpini opened the pull request:
Adds string to numeric conversion utility for shell.
Applies these functions to the Bluetooth Mesh Shell module.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

Andrewpini commented:
For any questions regarding if this belongs here on basis of its generic nature, ref: https://github.com/zephyrproject-rtos/zephyr/pull/43289

Andrewpini commented:
@alwa-nordic I know that you argued that this should not be moved back here, but I have spent way more time than intended for this task, and the discussion in the original PR does not seem to give a finite solution any time soon. I have cleared this with @carlescufi.

jhedberg reviewed (changes requested):
I really don’t think it makes sense to introduce such generic functionality in a mesh shell-only context. If necessary (to make progress introducing this as a generic API) we can discuss this in some appropriate Zephyr project meeting (like the API one).

ludvigsj commented on subsys/bluetooth/mesh/shell_util.c:
```diff
@@ -0,0 +1,202 @@
+	}
+
+	*val = temp_val;
+	return 0;
```
This can be simplified a lot. For one, `strtol()` wil skip leading whitespace completely on its own. It also convenientely returns the endptr for you if you provide it, which can be used for checking that only whitespace remains after the number. Combination of these makes this (and the unsigned version) simplify to
```c
char *endptr = NULL;

errno = 0;
*val = strtol(str, &endptr, 0);
if (errno == ERANGE) {
    return -ERANGE;
} else if (errno || endptr == NULL) {
    return -EINVAL;
}

/* Check that remaining string is whitespace.
Not sure if this is really needed now that this
is scoped to shell argument usage where
whitespace is already trimmed by the shell */
while ((char c = *endptr++)) {
    if (!isspace(c)) {
        return -EINVAL;
    }
}

return 0;
```

ludvigsj replied:
If we assume there should never be trailing whitespace, the while loop can be replaced with appending `|| *endptr != 0` to the `else if` above.

Andrewpini replied:
Thanks for the input 👍

ludvigsj commented on subsys/bluetooth/mesh/shell_util.c:
```diff
@@ -0,0 +1,202 @@
+	}
+
+	errno = 0;
+	temp_val = strtol(trimmed_buf, NULL, 10);
```
Why do we need to force decimal? What if I wanted to supply hex? I see no reason to force bases on users here.

Andrewpini replied:
Aggre with you that it could be nice to support hex, but as I showed you it creates some headaches for the strtodbl conversions and octal representations. I think decimal representation wil cover most usecases for now.

ludvigsj commented on subsys/bluetooth/mesh/shell_util.c:
```diff
@@ -0,0 +1,202 @@
+	double frac_dbl;
+	int err = 0;
+
+	size_t len = whitespace_trim(trimmed_buf, sizeof(trimmed_buf), str);
```
Is whitespace trimming really necessary anymore now that we live in the shell where whitespace is already trimmed?

Andrewpini replied:
Not 100% clear where this will end up, so keeping it until further notice.

ludvigsj commented on subsys/bluetooth/mesh/shell_util.c:
```diff
@@ -0,0 +1,202 @@
+{
+	long temp;
+
+	if (!strcmp(str, "on") || !strcmp(str, "enable")) {
```
Why these uncommon words but specifically not ``"true"``? Same goes for ``"false"`` below.

ludvigsj commented on subsys/bluetooth/mesh/shell_util.c:
```diff
@@ -0,0 +1,202 @@
+		frac_len = 9;
+	}
+
+	/* Avoid doing str2long if decimal part is empty */
```
The parts should be called "integer part" and "fractional part". The entire thing is decimal.

ludvigsj commented on subsys/bluetooth/mesh/shell_util.c:
```diff
@@ -0,0 +1,202 @@
+		return -EINVAL;
+	}
+
+	int comma_idx = strcspn(trimmed_buf, ".");
```
Could simplify a lot by using `strchr` here and saving to e.g. `char *frac_ptr`, as every use of `comma_idx` is either `trimmed_buf[comma_idx]` or `trimmed_buf + comma_idx`

Andrewpini replied:
Tried it, it comes at the cost of checking that `char *frack_ptr` is not NULL. I feel that the yield in simplicity is minimal whit this in mind.

ludvigsj commented on tests/bluetooth/mesh_shell_util/src/main.c:
```diff
@@ -0,0 +1,236 @@
+		  .exp_val.dbl_val.eps = 0.000000001f },
+		{ .is_valid = true, .str = " 58754.123    ", .exp_val.dbl_val.val = 58754.123,
+		  .exp_val.dbl_val.eps = 0.000000001f },
+		{ .is_valid = false, .str = "         " },
```
You're conveniently enough not testing the class of things like `"-12 . 3"` and `".      4"`, with spaces next to the decimal point. These would parse and ignore the internal whitespace. Not sure if that's on purpose.

Andrewpini replied:
(resolved in person) It would not return an unexpexted value, which is acceptable

ludvigsj commented on subsys/bluetooth/mesh/shell_util.c:
```diff
@@ -0,0 +1,189 @@
+		return -EINVAL;
+	}
+
+
```
Extra newline

PavelVPV commented on subsys/bluetooth/mesh/shell_util.c:
```diff
@@ -0,0 +1,189 @@
+#include <errno.h>
+#include "shell_util.h"
+
+static size_t whitespace_trim(char *out, size_t len, const char *str)
```
Won't whitespaces be already trimmed by shell module?

PavelVPV commented on subsys/bluetooth/mesh/shell_util.c:
```diff
@@ -0,0 +1,189 @@
+	return out_size;
+}
+
+int bt_mesh_str2l(const char *str, long *val)
```
```suggestion
int bt_mesh_str2long(const char *str, long *val)
```
or
```suggestion
int bt_mesh_strtol(const char *str, long *val)
```
`l` in some fonts looks like 1.

PavelVPV commented on subsys/bluetooth/mesh/shell_util.c:
```diff
@@ -0,0 +1,189 @@
+	char *endptr = NULL;
+	char c;
+
+	while (isspace(*str)) {
+		str++;
+	}
```
Won't whitespaces be already trimmed by shell module?

PavelVPV commented on subsys/bluetooth/mesh/shell_util.c:
```diff
@@ -0,0 +1,189 @@
+	return err;
+}
+
+int bt_mesh_str2bool(const char *str, bool *val)
```
Should this replace str2bool in `mesh/shell.c`? What about `str2u8`?

PavelVPV commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -3263,3 +3625,78 @@ static int cmd_mesh(const struct shell *shell, size_t argc, char **argv)
+	}
+
+
+	while ((c = *endptr++)) {
```
Is this needed?

Andrewpini commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -3263,3 +3625,78 @@ static int cmd_mesh(const struct shell *shell, size_t argc, char **argv)
+	while (isspace(*str)) {
+		str++;
+	}
+	if (*str == '-') {
```
No, me and @ludvigsj tested it a while ago, and the strtoul does not provide error on negative sign.

Andrewpini commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -3263,3 +3625,78 @@ static int cmd_mesh(const struct shell *shell, size_t argc, char **argv)
+		return 0;
+	}
+
+	while ((c = *endptr++)) {
```
No, if we assume that both heading and trailing whitespace is trimmed it is not neccesary

Andrewpini commented on subsys/bluetooth/mesh/shell.h:
```diff
@@ -0,0 +1,15 @@
+/* NOTE: None of these functions alters err reference upon success.
+ * Err param should be initialized to zero before use
+ */
+long bt_mesh_shell_str2l(const char *str, int base, int *err);
```
Sure 👍

Andrewpini commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -887,12 +943,16 @@ static int cmd_beacon_listen(const struct shell *shell, size_t argc,
-		uint8_t val = strtoul(argv[1], NULL, 0);
+		uint8_t val = bt_mesh_shell_str2ul(argv[1], 0, &err);
+		if (err) {
+			shell_warn(shell, "Unable to parse input string argument");
```
Well, I dont have a strong opinion on it. The only thing that entails a warning print is that the user wont get any feedback that the command did not execute. But I guess if most commands gives you some feedback on correct execution the user should understand that something went wrong? I agree that the print bloats the module, so it is a tradeoff situation.

ludvigsj replied:
I'm with Pavel here. Not being able to use these util functions without spamming the shell looks like a bad idea to me.

ludvigsj replied:
Removed comment, misunderstood this discussion

ludvigsj replied:
Returning error causes exactly nothing to happen when typing commands. I for one found this confusing when using the shell, as an empty response is not a good way to tell me I've typed something wrong.

PavelVPV replied:
Ok, I see. Then lets have a warning here.

PavelVPV commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -911,12 +973,19 @@ static int cmd_ttl(const struct shell *shell, size_t argc, char *argv[])
 		err = bt_mesh_cfg_friend_get(net.net_idx, net.dst, &frnd);
 	} else {
-		uint8_t val = str2u8(argv[1]);
+		uint8_t val = isdigit(argv[1][0]) ?
+				      bt_mesh_shell_strtoul(argv[1], 0, &err) :
+					    (uint8_t)bt_mesh_shell_strtobool(argv[1], 0, &err);
```
Why not just `bt_mesh_shell_strtobool`?

Andrewpini replied:
I decided to try to not alter the functionality of the existing implementation. Since `friend_set` technically has three valid friend states (`disabled`, `enabled`, `not supported`) strtobool won't fully cover these possibillities like the previous one. I think this was the reason why the previous str2u8 was implemented in a peculiar manner. IMO the `not supported` state is pretty useless for the set call, so I agree that it would be sufficient to use strtobool anyways.

Andrewpini replied:
The same goes for `relay set` and `gatt proxy set` aswell

PavelVPV replied:
I guess `not supported` can't be used in this case. It is a configuration message which can either enable or disable a feature.

Andrewpini replied:
Yes, I changed it for the three mentioned instances. I belive it was originally implemented like this because the spec does not clearly define which states are intended for set calls and which is only used in status responses.

PavelVPV commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -453,32 +466,24 @@ static int cmd_reset(const struct shell *shell, size_t argc, char *argv[])
 {
-	static bool enabled;
-	int err;
+	static bool enabled, onoff;
```
`onoff` should not be static?

PavelVPV commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -1341,10 +1452,15 @@ static int cmd_app_key_get(const struct shell *shell, size_t argc, char *argv[])
 	size_t cnt;
 	uint8_t status;
-	int err, i;
+	int err = 0, i;
```
Here and below. Split into 2 lines to avoid declaring initialized and non-initialized variables on the same line.

Andrewpini commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -3263,3 +3623,58 @@ static int cmd_mesh(const struct shell *shell, size_t argc, char **argv)
+		return false;
+	}
+
+	return (bool)bt_mesh_shell_strtol(str, base, err);
```
The reason for using `bt_mesh_shell_strtol` is to allow e.g. `-1` to be interpreted as a `true` conversion rather than it being evaluated as a invalid input. It is a corner case from a user perspective, but I do not see any particular reason why it should be prohibited. Do you?

Andrewpini replied:
I thought the conversion between bool and numeric required a cast. It apparently does not so i will remove it 👍

PavelVPV replied:
> The reason for using `bt_mesh_shell_strtol` is to allow e.g. `-1` to be interpreted as a `true` conversion rather than it being evaluated as a invalid input. It is a corner case from a user perspective, but I do not see any particular reason why it should be prohibited. Do you?

I'm not sure that `-1` should be interpreted as `true`.

Andrewpini replied:
Ok, I will change it

ludvigsj commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -3263,3 +3623,58 @@ static int cmd_mesh(const struct shell *shell, size_t argc, char **argv)
+	if (errno == ERANGE) {
+		*err = -ERANGE;
+		return 0;
+	} else if (errno || endptr == str || *endptr++) {
```
Don't need to increment endptr. Same below

jhedberg commented:
@jakub-uC the helpers presented in this PR are not in any way Mesh specific. Would it be possible to move them as part of the shell subsystem, so that any shell module can use them?

carlescufi commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -3263,3 +3623,58 @@ static int cmd_mesh(const struct shell *shell, size_t argc, char **argv)
 			cmd_mesh, 1, 1);
+
+
+long bt_mesh_shell_strtol(const char *str, int base, int *err)
```
Please move these 3 functions to `subsys/shell/shell_util.c`, that way other shell users can reuse them.

jhedberg commented:
You need to split the shell subsystem changes into a separate commit. Otherwise looks fine.

Andrewpini committed:
Shell: String to numeric conversion utils

Adds string to numeric conversion utility for shell.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

Andrewpini committed:
Bluetooth: Mesh: Add err check to shell input

Changes parsing of input string args  to provide error checking.
This is to prevent unintentional command execution on garbage input
strings.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>
